File: Lab - 02/Propuestos/01/ATMachine.java
public class ATMachine {
    private double balance;
    private static final double DEFAULT_INITIAL_BALANCE = 1000.0;
    
    /**
     * Constructor que inicializa el ATM con un saldo predeterminado
     */
    public ATMachine() {
        this.balance = DEFAULT_INITIAL_BALANCE;
    }
    
    /**
     * Constructor que permite establecer un saldo inicial personalizado
     * @param initialBalance El saldo inicial
     * @throws IllegalArgumentException si el saldo inicial es negativo
     */
    public ATMachine(double initialBalance) {
        if (initialBalance < 0) {
            throw new IllegalArgumentException("El saldo inicial no puede ser negativo");
        }
        this.balance = initialBalance;
    }
    
    /**
     * Procesa las operaciones del cajero automático
     * @param operationStr El código de operación (1-4)
     * @param amountStr El monto para depósito o retiro (opcional)
     * @return Un mensaje con el resultado de la operación
     */
    public String processOperation(String operationStr, String amountStr) {
        try {
            // Validar que la operación sea un número entero
            int operation = parseInteger(operationStr);
            
            switch (operation) {
                case 1: // Consultar Saldo
                    return checkBalance();
                    
                case 2: // Depositar Dinero
                    return deposit(amountStr);
                    
                case 3: // Retirar Dinero
                    return withdraw(amountStr);
                    
                case 4: // Salir
                    return "¡Gracias por utilizar nuestro cajero automático!";
                    
                default:
                    return "Error: Operación inválida. Por favor seleccione una opción válida (1-4).";
            }
        } catch (NumberFormatException e) {
            return "Error: " + e.getMessage();
        } catch (Exception e) {
            return "Error inesperado: " + e.getMessage();
        }
    }
    
    /**
     * Muestra el saldo actual
     * @return Un mensaje con el saldo actual
     */
    public String checkBalance() {
        return String.format("Su saldo actual es: S/.%.2f", balance);
    }
    
    /**
     * Realiza un depósito en la cuenta
     * @param amountStr El monto a depositar como cadena
     * @return Un mensaje con el resultado del depósito
     * @throws NumberFormatException si el monto no es un número válido
     */
    public String deposit(String amountStr) {
        try {
            double amount = parseAmount(amountStr);
            
            // Validar que el monto sea positivo
            if (amount <= 0) {
                return "Error: El monto a depositar debe ser un valor positivo.";
            }
            
            // Actualizar el saldo
            balance += amount;
            
            return String.format("Depósito exitoso de S/.%.2f\n%s", amount, checkBalance());
        } catch (NumberFormatException e) {
            return "Error: " + e.getMessage();
        }
    }
    
    /**
     * Realiza un retiro de la cuenta
     * @param amountStr El monto a retirar como cadena
     * @return Un mensaje con el resultado del retiro
     * @throws NumberFormatException si el monto no es un número válido
     */
    public String withdraw(String amountStr) {
        try {
            double amount = parseAmount(amountStr);
            
            // Validar que el monto sea positivo
            if (amount <= 0) {
                return "Error: El monto a retirar debe ser un valor positivo.";
            }
            
            // Validar que haya suficiente saldo
            if (amount > balance) {
                return String.format(
                    "Error: Saldo insuficiente. Su saldo actual es S/.%.2f y está intentando retirar S/.%.2f", 
                    balance, amount);
            }
            
            // Actualizar el saldo
            balance -= amount;
            
            return String.format("Retiro exitoso de S/.%.2f\n%s", amount, checkBalance());
        } catch (NumberFormatException e) {
            return "Error: " + e.getMessage();
        }
    }
    
    /**
     * Intenta parsear un String a un entero
     * @param value El valor a convertir
     * @return El valor como entero
     * @throws NumberFormatException si el valor no es un entero válido
     */
    private int parseInteger(String value) throws NumberFormatException {
        if (value == null || value.trim().isEmpty()) {
            throw new NumberFormatException("El valor de operación no puede estar vacío");
        }
        
        try {
            return Integer.parseInt(value.trim());
        } catch (NumberFormatException e) {
            throw new NumberFormatException("'" + value + "' no es una opción válida");
        }
    }
    
    /**
     * Intenta parsear un String a un monto (double)
     * @param value El valor a convertir
     * @return El valor como double
     * @throws NumberFormatException si el valor no es un número válido
     */
    private double parseAmount(String value) throws NumberFormatException {
        if (value == null || value.trim().isEmpty()) {
            throw new NumberFormatException("El monto no puede estar vacío");
        }
        
        // Reemplaza comas por puntos para manejar formatos numéricos diferentes
        String normalizedValue = value.trim().replace(',', '.');
        
        try {
            return Double.parseDouble(normalizedValue);
        } catch (NumberFormatException e) {
            throw new NumberFormatException("'" + value + "' no es un monto válido");
        }
    }
    
    /**
     * Obtiene el saldo actual (para pruebas)
     * @return El saldo actual
     */
    public double getBalance() {
        return balance;
    }
    
}

File: Lab - 02/Propuestos Python/Ej03/atmachine.py
class ATMachine:
    DEFAULT_INITIAL_BALANCE = 1000.0

    def __init__(self, initial_balance=None):
        if initial_balance is None:
            self.balance = ATMachine.DEFAULT_INITIAL_BALANCE
        else:
            if initial_balance < 0:
                raise ValueError("El saldo inicial no puede ser negativo")
            self.balance = initial_balance

    def process_operation(self, operation_str, amount_str=None):
        try:
            operation = self._parse_integer(operation_str)

            if operation == 1:
                return self.check_balance()
            elif operation == 2:
                return self.deposit(amount_str)
            elif operation == 3:
                return self.withdraw(amount_str)
            elif operation == 4:
                return "¡Gracias por utilizar nuestro cajero automático!"
            else:
                return "Error: Operación inválida. Por favor seleccione una opción válida (1-4)."
        except ValueError as e:
            return f"Error: {str(e)}"
        except Exception as e:
            return f"Error inesperado: {str(e)}"

    def check_balance(self):
        return f"Su saldo actual es: S/.{self.balance:.2f}"

    def deposit(self, amount_str):
        try:
            amount = self._parse_amount(amount_str)
            if amount <= 0:
                return "Error: El monto a depositar debe ser un valor positivo."
            self.balance += amount
            return f"Depósito exitoso de S/.{amount:.2f}\n{self.check_balance()}"
        except ValueError as e:
            return f"Error: {str(e)}"

    def withdraw(self, amount_str):
        try:
            amount = self._parse_amount(amount_str)
            if amount <= 0:
                return "Error: El monto a retirar debe ser un valor positivo."
            if amount > self.balance:
                return (f"Error: Saldo insuficiente. Su saldo actual es S/.{self.balance:.2f} "
                        f"y está intentando retirar S/.{amount:.2f}")
            self.balance -= amount
            return f"Retiro exitoso de S/.{amount:.2f}\n{self.check_balance()}"
        except ValueError as e:
            return f"Error: {str(e)}"

    def _parse_integer(self, value):
        if value is None or value.strip() == "":
            raise ValueError("El valor de operación no puede estar vacío")
        try:
            return int(value.strip())
        except ValueError:
            raise ValueError(f"'{value}' no es una opción válida")

    def _parse_amount(self, value):
        if value is None or value.strip() == "":
            raise ValueError("El monto no puede estar vacío")
        normalized_value = value.strip().replace(",", ".")
        try:
            return float(normalized_value)
        except ValueError:
            raise ValueError(f"'{value}' no es un monto válido")

    def get_balance(self):
        return self.balance